Fixed: Chinese Anime releases that include a season number

zx900930 opened the pull request:
#### Database Migration
NO

#### Description
Updated: 2013/01/18

This fix is a continuation to https://github.com/Sonarr/Sonarr/pull/5356

After the patch https://github.com/Sonarr/Sonarr/pull/5356
Sonarr can correctly grab
`[Lilith-Raws] 致不灭的你 / Anime-Series Title S02 - 12 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]`
as `Season 2`. 
But when importing them from download clients, sonarr still imported it to `Season 1`
![image](https://user-images.githubusercontent.com/734749/213301478-1144aedb-5f19-45e4-8401-e6a1e79b3146.png)

The root cause of this type of bug:

1. Sonarr will parse 2 different names when downloading series.
    One is release name, in this case: `[Lilith-Raws] 致不灭的你 / Anime-Series Title S02 - 12 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]`.
    The other one is downloaded file name, in this case: `[Lilith-Raws] Anime-Series Title S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4].mp4`.
2. XEM normaly doesn't have `S02` alias.
3. The patch https://github.com/Sonarr/Sonarr/pull/5356 make sonarr passed the step 1: send that release to download client    as `S02E12`.
    But after file was downloaded, the import process was failed because filename `[Lilith-Raws] Anime-Series Title S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4].mp4` get parsed and `Anime-Series Title S02` was not found on XEM so it get imported directly into `Season 1`

This PR can resolve this issue by simply replacing the `S0x` by `Sx` if it matches below rules:
1. Release group must have chinese characters or in one of these `LoliHouse`|`Lilith-Raws`|`Skymoon-Raws`|`orion origin`.
2. The `filename`/`release name` should in `title S0x - abs` format

thezoggy commented:
we've had plenty of people ask for chinese aliases to be added to xem and all the ones ive seen requested are for releases that do absolute numbering...    the only issue with chinese alias requests is when they use a non standard format which causes them to be parsed funky so the series title parsed is not the expected part of the release.

zx900930 commented:
> we've had plenty of people ask for chinese aliases to be added to xem and all the ones ive seen requested are for releases that do absolute numbering... the only issue with chinese alias requests is when they use a non standard format which causes them to be parsed funky so the series title parsed is not the expected part of the release.

Yep, but the problem is even if we added chinese aliases, because the release groups are using absolute numbering scheme.

Those release will still hit one of below regex and causing import error
```
// Anime - [SubGroup] Title with trailing number Absolute Episode Number - Batch separated with tilde
                new Regex(@"^\[(?<subgroup>.+?)\][-_. ]?(?<title>.+?[^-]+?)(?:(?<![-_. ]|\b[0]\d+) - )[-_. ]?(?<absoluteepisode>\d{2,3}(\.\d{1,2})?(?!\d+))\s?~\s?(?<absoluteepisode>\d{2,3}(\.\d{1,2})?(?!\d+))(?:[-_. ]+(?<special>special|ova|ovd))?.*?(?<hash>\[\w{8}\])?(?:$|\.mkv)",
                    RegexOptions.IgnoreCase | RegexOptions.Compiled),

// Anime - [SubGroup] Title with trailing number Absolute Episode Number
                new Regex(@"^\[(?<subgroup>.+?)\][-_. ]?(?<title>[^-]+?)(?:(?<![-_. ]|\b[0]\d+) - )(?:[-_. ]?(?<absoluteepisode>\d{2,3}(\.\d{1,2})?(?!\d+)))+(?:[-_. ]+(?<special>special|ova|ovd))?.*?(?<hash>\[\w{8}\])?(?:$|\.mkv)",
                          RegexOptions.IgnoreCase | RegexOptions.Compiled),

// Anime - [SubGroup] Title with trailing number Absolute Episode Number
                new Regex(@"^\[(?<subgroup>.+?)\][-_. ]?(?<title>[^-]+?)(?:(?<![-_. ]|\b[0]\d+)[_ ]+)(?:[-_. ]?(?<absoluteepisode>\d{3}(\.\d{1,2})?(?!\d+|-[a-z]+)))+(?:[-_. ]+(?<special>special|ova|ovd))?.*?(?<hash>\[\w{8}\])?(?:$|\.mkv)",
                          RegexOptions.IgnoreCase | RegexOptions.Compiled),

// Anime - [SubGroup] Title - Absolute Episode Number
new Regex(@"^\[(?<subgroup>.+?)\][-_. ]?(?<title>.+?)(?:(?<!\b[0]\d+))(?:[. ]-[. ](?<absoluteepisode>\d{2,3}(\.\d{1,2})?(?!\d+|[-])))+(?:[-_. ]+(?<special>special|ova|ovd))?.*?(?<hash>\[\w{8}\])?(?:$|\.mkv)",
                          RegexOptions.IgnoreCase | RegexOptions.Compiled),
```

The methods I've tryed:
1. Adding a special mark to `PreSubstitutionRegex` output strings like `#sonarr_chinese_anime#` and followed by a subsitution regex dealing with it.
-> Failed, it will cause false positive and the false positive can't be parsed due to added mark.

2. Adding an `E` before the pharsed episode number.
-> Same as above, will affect false positives that do use absolute numbering. The difficulty is using regex to determine if it was an chinese anime release, becuase some english releases will add original japanese title, sharing the same unicode range `\u4E00-\u9FCC` causing false positives.

3. Like this PR, add additional rules to `PreSubstitutionRegex` , normalize the string for certain release groups.
-> It works and non destructive, but the limitation is can't take advantage of xem mapping to get the different allias for season number.

Some solutions I've thinking:
1. Let user create their own rules in `Custom Formats` that can formating titles from their preferd release groups to a normalized sonarr format. 

2. Add a indexer language tag to the titles, if a release came from a chinese anime indexer, apply chinese anime `SubstitutionRegex`.

thezoggy commented:
how other apps generally have handled this is have the language specific fork, where the said fork have their own rulesets/qualities/regexes/etc. this of course then is usually maintained by said fork and not part of the actual app to prevent the collision.   

how sonarr might achieve this is to have it dynamically just load said <language>-regex file if found, otherwise the system defaults is just used. of course this is a bit of complexity to be had for niche usage

markus101 commented on src/NzbDrone.Core.Test/ParserTests/UnicodeReleaseParserFixture.cs:
```diff
@@ -47,7 +47,35 @@ public void should_parse_chinese_anime_releases(string postTitle, string title,
         [TestCase("[LoliHouse] Love Live! 虹咲学园学园偶像同好会 第二季 / Anime-Series Title S2 - 10 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]", "Anime-Series Title", "LoliHouse", 2, 10)]
         [TestCase("[澄空学园&雪飘工作室&LoliHouse] 辉夜大小姐想让我告白 第三季 / Anime-Series Title S3 - 06 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]", "Anime-Series Title", "澄空学园&雪飘工作室&LoliHouse", 3, 6)]
-        public void should_parse_chinese_anime_season_episode_releases(string postTitle, string title, string subgroup, int seasonNumber, int episodeNumber)
+        public void should_parse_chinese_anime_season_episode_releases_1(string postTitle, string title, string subgroup, int seasonNumber, int episodeNumber)
```
These can all be in one fixture. `should_parse_chinese_anime_season_number_with_episode_number_releases`

markus101 commented on src/NzbDrone.Core/Parser/Parser.cs:
```diff
@@ -25,15 +25,21 @@ public static class Parser
                 new RegexReplace(@"^\[(?:(?<subgroup>[^\]]+?)(?:[\u4E00-\u9FCC]+)?)\]\[(?<title>[^\]]+?)(?:\s(?<chinesetitle>[\u4E00-\u9FCC][^\]]*?))\]\[(?:(?:[\u4E00-\u9FCC]+?)?(?<episode>\d{1,4})(?:[\u4E00-\u9FCC]+?)?)\]", "[${subgroup}] ${title} - ${episode} - ", RegexOptions.Compiled),
 
-                // Chinese LoliHouse/Lilith-Raws releases use Season + Absolute Episode Number, normalize using SxxExx format
+                // Chinese Anime Season Episode Rule 1: Chinese LoliHouse/Lilith-Raws releases use Season - Episode Number format, normalize using SxxExx format
```
Remove the `Rule #`. `Chinese Anime with Season and Episode numbers: ...`

zx900930 replied:
> How do these releases differ from other anime that uses `Some title S2 - 01` for the release? With XEM aliases those are treated as season 2 aliases with absolute numbers that reset at the beginning of the season, not unlike season/episode numbers.
> 
> I believe @Taloth opened an issue for user provided substitution regexes, but how that would be achieved has not been fleshed out.

Without these 2 new regex rules:
![image](https://user-images.githubusercontent.com/734749/212018857-12ecf81b-8283-46bc-b19f-23583143fa6b.png)
![image](https://user-images.githubusercontent.com/734749/212019058-39c934e6-6b63-4885-878b-be423ebc69d6.png)



The problem is: 
The xem doesn't have `Some title S02` as an alias, the `SubstitutionRegex` is dealing with this situation to convert  `Some title S02` to xem mapped  `Some title S2` title, but for chinese releases, because they are pre-formatted with `PreSubstitutionRegex` to extract the English titles in it, it won't catch by those regex rules that formatting the `S02` to `S2`,  so the relase name will become `Some title S02 - 01` or `Some title S02 [01]` and the 01 is treated as absolute episodes number 01 -> `S1E1`

Just give you an idea how S02 is popular as an naming scheme in chinese releases, I don't know if xem can batch edit every anime to add S02 S03 S04 alias, IMO using regex is less painful than added S0x alias to every animes in xem.
![image](https://user-images.githubusercontent.com/734749/212020586-a284922f-4ab4-4619-aac5-677941c8b133.png)

markus101 reviewed (commented):
How do these releases differ from other anime that uses `Some title S2 - 01` for the release? With XEM aliases those are treated as season 2 aliases with absolute numbers that reset at the beginning of the season, not unlike season/episode numbers.

I believe @Taloth opened an issue for user provided substitution regexes, but how that would be achieved has not been fleshed out.

thezoggy commented:
from what i can tell this user is confused on what absolute numbering is and season alises. they are reading the release as s#e# not what it actually is, series - abs. so if this sort of change got added, it would break many xem mapping and parsing of things

zx900930 commented:
> from what i can tell this user is confused on what absolute numbering is and season aliases. they are reading the release as s# - e# not what it actually is, series - abs. i can see that they are trying to solve the need of having season aliases so part of the series title gets used as the season and then the abs number of the season is treated as the episode.. and while this may work for some scenarios i can see this breaking many things including any existing xem mapping for the series that has the season aliases

Thanks for your advice, I found a much straight forward way:

If I simply remove the 0 from the S02 to make it to S2 and don't touch the abs numbering part, then it won't break anything right?

zx900930 commented:
After discussion with @thezoggy on discord, I'm using this new simple approach:
If the release string is in `title S0x - abs` format then remove the first `0` right after the `S` to convert it to `title Sx - abs` format.
Thus it can take full advantage of XEM that already has a Sx in the aliases.
And it won't break any english releases using `title Sxx - abs`  format.

To ensure it won't break previously working releases that has S02 alias in it, I'v tested the new patch with 
https://thexem.info/xem/show/5460

Before Patch:
```
2023-01-13 02:22:08.9|Debug|Parser|Parsing string '[Lilith-Raws] Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4].mp4'
2023-01-13 02:22:08.9|Trace|Parser|Replace regex: ^\[(?<subgroup>[^\]]*?(?:LoliHouse|ZERO|Lilith-Raws)[^\]]*?)\](?<title>[^\[\]]+?)(?: - (?<episode>[0-9-]+)\s*|\[第?(?<episode>[0-9]+(?:-[0-9]+)?)话?(?:END|完)?\])\[
2023-01-13 02:22:08.9|Debug|Parser|Substituted with [Lilith-Raws][ Muv-Luv Alternative S02][11][Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]
2023-01-13 02:22:08.9|Trace|Parser|Replace regex: ^\[(?<subgroup>[^\]]+)\](?:\s?★[^\[ -]+\s?)?\[?(?:(?<chinesetitle>(?=[^\]]*?[\u4E00-\u9FCC])[^\]]*?)(?:\]\[|\s*[_/·]\s*)){0,2}(?<title>[^\]]+?)\]?(?:\[\d{4}\])?\[第?(?<episode>[0-9]+(?:-[0-9]+)?)(?:话|集)?(?: ?END|完| ?Fin)?\]
2023-01-13 02:22:08.9|Debug|Parser|Substituted with [Lilith-Raws]  Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]
2023-01-13 02:22:08.9|Trace|Parser|^\[(?<subgroup>.+?)\][-_. ]?(?<title>.+?)(?:(?<!\b[0]\d+))(?:[. ]-[. ](?<absoluteepisode>\d{2,3}(\.\d{1,2})?(?!\d+|[-])))+(?:[-_. ]+(?<special>special|ova|ovd))?.*?(?<hash>\[\w{8}\])?(?:$|\.mkv)
2023-01-13 02:22:08.9|Debug|Parser|Episode Parsed. Muv-Luv Alternative S02 - 011 
2023-01-13 02:22:08.9|Debug|Parser|Languages parsed: Chinese
2023-01-13 02:22:08.9|Debug|QualityParser|Trying to parse quality for [Lilith-Raws] Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4].mp4
2023-01-13 02:22:08.9|Debug|Parser|Quality parsed: WEBDL-1080p v1
2023-01-13 02:22:08.9|Debug|Parser|Release Group parsed: Lilith-Raws
2023-01-13 02:22:08.9|Debug|Parser|Parsing string '[Lilith-Raws]  Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]'
2023-01-13 02:22:08.9|Debug|SceneMappingService|Found scene mapping for: Muv-Luv Alternative S02. TVDB ID for mapping: 402231
2023-01-13 02:22:08.9|Debug|Parser|Parsing string '[Lilith-Raws]  Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]'
2023-01-13 02:22:08.9|Debug|SceneMappingService|Found scene mapping for: Muv-Luv Alternative S02. TVDB ID for mapping: 402231
2023-01-13 02:22:08.9|Debug|ParsingService|Using absolute episode number 11 for: Muv-Luv Alternative - TVDB: 2x11
```

After Patch:
```
2023-01-13 02:15:46.2 Debug: Parsing string '[Lilith-Raws] Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4].mp4' 
2023-01-13 02:15:46.2 Trace: Replace regex: ^\[(?<subgroup>[^\]]*?(?:LoliHouse|ZERO|Lilith-Raws)[^\]]*?)\](?<title>[^\[\]]+?)(?: - (?<episode>[0-9-]+)\s*|\[第?(?<episode>[0-9]+(?:-[0-9]+)?)话?(?:END|完)?\])\[ 
2023-01-13 02:15:46.2 Debug: Substituted with [Lilith-Raws][ Muv-Luv Alternative S02][11][Baha][WEB-DL][1080p][AVC AAC][CHT][MP4] 
2023-01-13 02:15:46.2 Trace: Replace regex: ^\[(?<subgroup>[^\]]+)\](?:\s?★[^\[ -]+\s?)?\[?(?:(?<chinesetitle>(?=[^\]]*?[\u4E00-\u9FCC])[^\]]*?)(?:\]\[|\s*[_/·]\s*)){0,2}(?<title>[^\[\]]+?)(?:\s(?:S?(?<!\d+)((0)(?<season>\d)|(?<season>[1-9]\d))(?!\d+)))\]?(?:\[\d{4}\])?\[第?(?<episode>[0-9]+(?:-[0-9]+)?)(?:话|集)?(?: ?END|完| ?Fin)?\] 
2023-01-13 02:15:46.2 Debug: Substituted with [Lilith-Raws]  Muv-Luv Alternative S2 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4] 
2023-01-13 02:15:46.2 Trace: ^\[(?<subgroup>.+?)\][-_. ]?(?<title>.+?)(?:(?<!\b[0]\d+))(?:[. ]-[. ](?<absoluteepisode>\d{2,3}(\.\d{1,2})?(?!\d+|[-])))+(?:[-_. ]+(?<special>special|ova|ovd))?.*?(?<hash>\[\w{8}\])?(?:$|\.mkv) 
2023-01-13 02:15:46.2 Debug: Episode Parsed. Muv-Luv Alternative S2 - 011  
2023-01-13 02:15:46.2 Debug: Languages parsed: Chinese 
2023-01-13 02:15:46.2 Debug: Trying to parse quality for [Lilith-Raws] Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4].mp4
2023-01-13 02:15:46.2 Debug: Quality parsed: WEBDL-1080p v1 
2023-01-13 02:15:46.2 Debug: Release Group parsed: Lilith-Raws
2023-01-13 02:15:46.2|Debug|Parser|Parsing string '[Lilith-Raws]  Muv-Luv Alternative S02 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]'
2023-01-13 02:15:46.2|Debug|SceneMappingService|Found scene mapping for: Muv-Luv Alternative S2. TVDB ID for mapping: 402231
2023-01-13 02:15:46.2|Debug|Parser|Parsing string '[Lilith-Raws]  Muv-Luv Alternative S2 - 11 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]'
2023-01-13 02:15:46.2|Debug|SceneMappingService|Found scene mapping for: Muv-Luv Alternative S2. TVDB ID for mapping: 402231
2023-01-13 02:15:46.2|Debug|ParsingService|Using absolute episode number 11 for: Muv-Luv Alternative - TVDB: 2x11
```
You can see before patch it was using the S02 alias and after patch it was using the S2 alias, both of them are successfully imported as 2x11.

zx900930 commented:
Improvements planned 2023/01/18:
Added support for chinese anime release groups that use `S0x` as season number and don't have any CJK characters in the group name. Using strict group name match so won't create false positives.

added release groups:
Lilith-Raws/Skymoon-Raws/orion origin/UHA-WINGS

zx900930 commented:
some test results with this new patch
![image](https://user-images.githubusercontent.com/734749/213207515-40848ef9-a08a-44bc-8c24-10b98042cfdb.png)

markus101 commented on src/NzbDrone.Core.Test/ParserTests/UnicodeReleaseParserFixture.cs:
```diff
@@ -3,10 +3,12 @@
 using NUnit.Framework;
 using NzbDrone.Core.Indexers;
 using NzbDrone.Core.Test.Framework;
+using Sentry;
```
This shouldn't be needed in tests.

zx900930 replied:
removed

markus101 commented on src/NzbDrone.Core.Test/ParserTests/UnicodeReleaseParserFixture.cs:
```diff
@@ -3,10 +3,12 @@
 using System.Linq;
 using System.Text;
 using System.Threading.Tasks;
+using FizzWare.NBuilder.Dates;
```
Is this needed?

zx900930 replied:
no, it's not used.

markus101 reviewed (changes requested):
Looks like a couple unnecessary namespaces are being included, but otherwise this is looking good.
